# :dna: DNA replication dynamics in ecDNA :dna:

This GitHub repository contains the analysis code for the paper _ecDNA replication is disorganised and vulnerable to replication stress_ which is available on [BioRxiv](https://www.biorxiv.org/content/10.1101/2025.03.22.644567v1.full.pdf).

This project is a collaboration between [Julian Sale Lab](https://www2.mrc-lmb.cam.ac.uk/groups/jes/) and [Boemo Group](https://www.boemogroup.org)

#  :gear: DNAscent analysis pipeline

We followed the workflow for DNAscent version 3.1.2 outlined in the [DNAscent documentation](https://dnascent.readthedocs.io/en/latest/workflows.html).

### :chart_with_upwards_trend: 1. Guppy (basecalling)

Raw signals (.fast5 format) were basecalled using the Oxford Nanopore Technologies Guppy Basecalling Software (version 5.0.16). We specified the PromethION configuration file for DNA nanopore version R9.4.1 with `dna_r9.4.1_450bps_hac_prom.cfg`

### :round_pushpin: 2. Minimap2 (alignment)
Basecalled reads (both failed and passed) were merged into one *.fastq* file and aligned to the human reference genome hg38 `GCF_000001405.40_GRCh38.p14_genomic.fna` using the Minimap2 software (version 2.24). The aligned *.sam* output file was converted to a *.bam* file using samtools (version 1.14), sorted and indexed.

### :card_file_box: 3. DNAscent index

DNAscent index (version 3.1.2) was run on the raw singnal *.fast5* files alongside the `sequencing_summary.txt` file generated by Guppy.

### :mag_right: 4. DNAscent detect

DNAscent detect (version 3.1.2) was run using `-l 20000 -q 20` specifications.

### :arrow_right: 5. DNAscent forkSense

DNAscent forkSense (version 3.1.2) was run using `--markOrigins --markTerminations --markForks --markAnalogues --makeSignatures --order EdU,BrdU` specifications

# :bar_chart: Visualisations

### :open_file_folder: 6. forkSense to csv conversion

The forkSense output `.bed` files were converted to a `.csv` file (using a custom [python script](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/01_DNAscent_csv/createCsv_forks.py))and fork speed was calculated for all forks that were >= 2000 bp from the end of mapped read, to avoid understimation of fork speed for forks where the fork ran over the end of the mapped read. For fork speed calculation, we used a pulse time of 12 minutes (6 minutes EdU, 6 minutes BrdU).

**left/right forks: `createCsv_forks.py` file names after processing DNAscent forkSense output for hg38 or ecDNA mapped forks:**

> - DM_ecDNA = '20240405_directories_DM_ecDNA.csv'
> - DM_hg38 = '20240405_directories_DM_hg38.csv'
> - HSR_ecDNA = '20240405_directories_HSR_ecDNA.csv'
> - HSR_hg38 = '20240405_directories_HSR_hg38.csv'
> - DM_ecDNA_HU = '20240405_directories_DM_ecDNA_HU.csv'
> - DM_hg38_HU = '20240405_directories_DM_hg38_HU.csv'


**origins files: `createCsv_origins.py` :**

> - ori_DM_ecDNA = 20240503_directories_DM_ecDNA_origins.csv
> - ori_DM_hg38 = 20240503_directories_DM_hg38_origins.csv
> - ori_HSR_ecDNA = 20240503_directories_HSR_ecDNA_origins.csv
> - ori_HSR_hg38 = 20240503_directories_HSR_hg38_origins.csv
> - ori_DM_ecDNA_HU = 20240503_directories_DM_ecDNA_HU_origins.csv
> - ori_DM_hg38_HU = 20240503_directories_DM_hg38_HU_origins.csv

### :bar_chart: 7. Visualisation and statistics

Replication fork speed, stall scores, origin GC content and origin firing alongside other statistics were calculated using Python. The [Jupyter notebook](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/02_Jupyter_notebook/analysis.ipynb) contains all visualisation and statistics steps for all main and supplementary figures except for circular plots which were created in R (see section 8).

### :large_blue_circle: 8. Circular ecDNA visualisations

We used the [circlize package](https://doi.org/10.1093/bioinformatics/btu393) for which a tutorial is available [here](https://jokergoo.github.io/circlize_book/book/). DNAscent replication fork speeds, stall scores and positions of origins of replicaiton were derived from the `.csv` files generated in **step 6**. External data including SNS-seq and Ini-seq origin locations were taken from the files listed in the table below:


| Element            | Source        | File       |
|--------------------|---------------|------------|
| origins            | DNAscent ([custom python script](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/01_DNAscent_csv/createCsv_origins.py))    | 20240503_directories_XX_ecDNA_origins.csv |
| forks & stall scores | DNAscent ([custom python script](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/01_DNAscent_csv/createCsv_forks.py))   | 20240531_directories_df_XX_ecDNA_HU_iqr_filtered.csv |
| read coverage      | Minimap2 `.bam` | 202410017_XX_ecDNA_coverage_bam_filtered.csv |
| SNS-seq origins    | Akerman 2020  | [Table1a_Origin_coordinates in Supplementary Data 1](https://www.nature.com/articles/s41467-020-18527-0#Sec33) |
| Ini-seq origins    | Guilbaud 2022 | [GSM5658908_Ini-seq2.called.replication.origins.bed](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM5658908) |
| GC-content         | ecDNA reference map & [Python script](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/02_Jupyter_notebook/gc_content_ecDNA.py) `gc_content_ecDNA.py` | 20240529_gc_content_5000_bp.csv |
| G4 scores          | [G4 Hunter](https://academic.oup.com/nar/article/44/4/1746/1854457) | [input file](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/00_resources/basic_ecDNAseq_indexfile_formatted.fasta) |
| gene locations     | [hg38 genome assembly](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/) | GCF_000001405.40_GRCh38.p14_genomic.fna |

The custom ecDNA reference map backbone for the circlize package was created using a [custom Python script](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/02_Jupyter_notebook/create_cytoband_file.py) `create_cytoband_file.py` based on the ecDNA reference map `basic_ecDNAseq_indexfile_formatted.fasta` from AmpliconArchitect and is available as `basic_ecDNAseq_indexfile_formatted_cytoband.txt`.

For origin locations, replication fork speed, stall scores and fork orientation, the interquartile range (IQR) filtered `.csv`files created in the [Jupyter notebook](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/02_Jupyter_notebook/analysis.ipynb) were used as input. XX in the file name represents either DM, HSR or DM_HU.

G4 scores for the ecDNA reference map were calculated using [G4 Hunter](https://academic.oup.com/nar/article/44/4/1746/1854457) which has a [web application](https://academic.oup.com/bioinformatics/article/35/18/3493/5306941) and Python code to run G4Hunter is available on [GitHub](https://github.com/AnimaTardeb/G4Hunter.git). We ran the Python script with parameters `-w 25 -s 1.0`. We then filtered for G4 calls using a threshold of >= 1.5.

GC content was averaged for each 5000 bp (5 kb) segment to obtain a resolution that is easily visible on the ecDNA reference map.

The read coverage was obtained by extracting reads that passed the minimum DNAscent quality criteria (mapping quality >= 20 and aligned read length >= 20000 using a custom Python script `coverage_bam_ecDNA.py` or `coverage_bam_hg38.py` to process the Minimap2 aligned `.bam`files. XX in the read coverage csv file name represents either DM, HSR or DM_HU.


### :snake: Python version

Steps 6, 7 and 8 were performed using Python version 3.11.5 in a virtual environment, all required packages for installation are listed in the `requirements.txt` [file](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/requirements.txt).


### :desktop_computer: R version

All circular plots were created using R version 4.3.3 in a R Markdown script which is available [here](https://github.com/Pfuderer/ecDNA_replication_dynamics/blob/main/03_R_circos_plots/ecDNA_replication.md). The file also has a list with all package versions at the end of the script.

# :paperclip: Citation

Please cite our paper when using any of the code in this repository and reference this GitHub repository.

> Jaworski, J. J., Pfuderer, P. L., Czyz, P., Petris, G., Boemo, M. A., & Sale, J. (2025). ecDNA replication is disorganised and vulnerable to replication stress. bioRxiv, 2025-03.

# :mailbox: Contact

Please reach out to Pauline via email with any questions plp27[at]cam.ac.uk.
